src/platform/interpreter/reservedNamedProvider.node.ts

```diff
@@ -0,0 +1,146 @@
+        );
+        workspace.onDidChangeConfiguration(
+            (e) => {
+                if (e.affectsConfiguration(`jupyter.${ignoreListSettingName}`)) {
+                    this.initializeIgnoreList();
+                }
+            },
+            this,
+            this.disposables
+        );
+        this.initializeIgnoreList();
+    }
+
+    public dispose() {
+        disposeAllDisposables(this.disposables);
+    }
+
+    public async getUriOverridingReservedPythonNames(cwd: Uri): Promise<{ uri: Uri; type: 'file' | '__init__' }[]> {
+        const [files, initFile] = await Promise.all([
+            this.fs.searchLocal('*.py', cwd.fsPath, true),
+            // Look for packages in the current directory (only interested in top level).
+            // Just as xml.py could end up overriding the built in xml module, so can xml/__init__.py.
+            this.fs.searchLocal('*/__init__.py', cwd.fsPath, true)
+        ]);
+        const problematicFiles: { uri: Uri; type: 'file' | '__init__' }[] = [];
+        const filePromises = Promise.all(
+            files
+                .map((file) => Uri.joinPath(cwd, file))
+                .map(async (uri) => {
+                    if (await this.isReserved(uri)) {
+                        problematicFiles.push({ uri, type: 'file' });
+                    }
+                })
+        );
+        const initFilePromises = Promise.all(
+            initFile
+                .map((file) => Uri.joinPath(cwd, file))
+                .map(async (uri) => {
+                    if (await this.isReserved(uri)) {
+                        problematicFiles.push({ uri, type: '__init__' });
+                    }
+                })
+        );
+        await Promise.all([filePromises, initFilePromises]);
+        return problematicFiles;
+    }
+    public async isReserved(uri: Uri): Promise<boolean> {
+        // Lets keep it simple and focus only on plain text python files.
+        if (!uri.fsPath.toLowerCase().endsWith('.py')) {
+            return false;
+        }
+
+        await this.pendingUpdate;
+        const filePath = this.platform.isWindows ? uri.fsPath.toLowerCase() : uri.fsPath;
+        if (
+            Array.from(this.ignoredFiles).some((item) => {
+                if (item === filePath || minimatch(filePath, item, { dot: true })) {
+                    return true;
+                }
+            })
+        ) {
+            return false;
+        }
+
+        // Use the name of the file as the module name.
+        const baseName = path.basename(uri, path.extname(uri)).toLowerCase();
+        // If its a __init__.py, get name of parent folder (as its a module).
+        const possibleModule = baseName === '__init__' ? path.basename(path.dirname(uri)).toLowerCase() : baseName;
+        if (this.cachedModules.has(possibleModule)) {
+            return true;
+        }
+
+        if (this.extensionChecker.isPythonExtensionInstalled) {
```

Comment:
Would this check make sense inside of package.listPackages? Thinking about the case that I was looking at (which you've resolved in this PR I believe) where this was getting pulled in with telemetry. Someone consuming listPackage might not be aware that they are creating an install prompt for the PythonExtension when using it.

Reply:
Good point